models/src/main/scala/coop/rchain/models/rholangn/parmanager/Serialization.scala

```diff
@@ -2,256 +2,177 @@ package coop.rchain.models.rholangn.parmanager
-            write(send.persistent)
-
-        case receive: ReceiveN =>
-          write(RECEIVE) *>
-            writeSeq(receive.sortedBinds) *>
-            writePar(receive.body) *>
-            write(receive.persistent) *>
-            write(receive.peek) *>
-            write(receive.bindCount)
-
-        case m: MatchN => write(MATCH) *> writePar(m.target) *> writeSeq(m.cases)
-
-        case n: NewN =>
-          write(NEW) *>
-            write(n.bindCount) *>
-            writePar(n.p) *>
-            writeSeq[String](n.sortedUri, write) *>
-            writeSeq[(String, ParN)](n.sortedInjections, writeTupleStringPar)
-
-        /* Collections */
-        case eList: EListN   => write(ELIST) *> writeSeq(eList.ps) *> writeOpt(eList.remainder)
-        case eTuple: ETupleN => write(ETUPLE) *> writeSeq(eTuple.ps)
-        case eSet: ESetN     => write(ESET) *> writeSeq(eSet.sortedPs) *> writeOpt(eSet.remainder)
-        case eMap: EMapN =>
-          write(EMAP) *>
-            writeSeq[(ParN, ParN)](eMap.sortedPs, writeTuplePar) *>
-            writeOpt(eMap.remainder)
-
-        /* Connective */
-        case connAnd: ConnAndN => write(CONNECTIVE_AND) *> writeSeq(connAnd.ps)
-        case connOr: ConnOrN   => write(CONNECTIVE_OR) *> writeSeq(connOr.ps)
-
-        case eMethod: EMethodN =>
-          write(EMETHOD) *>
-            write(eMethod.methodName) *>
-            writePar(eMethod.target) *>
-            writeSeq(eMethod.arguments)
-
-        /* Auxiliary types */
-
-        case bind: ReceiveBindN =>
-          write(RECEIVE_BIND) *>
-            writeSeq(bind.patterns) *>
-            writePar(bind.source) *>
-            writeOpt(bind.remainder) *>
-            write(bind.freeCount)
-
-        case mCase: MatchCaseN =>
-          write(MATCH_CASE) *>
-            writePar(mCase.pattern) *>
-            writePar(mCase.source) *>
-            write(mCase.freeCount)
-
-        case unknownType => throw new Exception(s"Unknown type `$unknownType`")
-      }
-    }
+      case receive: ReceiveN =>
+        write(RECEIVE) *>
+          writeSeq(receive.sortedBinds, writePar) *>
+          writePar(receive.body) *>
+          write(receive.persistent) *>
+          write(receive.peek) *>
+          write(receive.bindCount)
+
+      case m: MatchN => write(MATCH) *> writePar(m.target) *> writeSeq(m.cases, writePar)
+
+      case n: NewN =>
+        write(NEW) *>
+          write(n.bindCount) *>
+          writePar(n.p) *>
+          writeSeq[String](n.sortedUri, write) *>
+          writeSeq[(String, ParN)](n.sortedInjections, writeTupleStringT(_, writePar))
+
+      /* Collections */
+      case eList: EListN =>
+        write(ELIST) *> writeSeq(eList.ps, writePar) *> writeOpt(eList.remainder, writePar)
+      case eTuple: ETupleN => write(ETUPLE) *> writeSeq(eTuple.ps, writePar)
+      case eSet: ESetN =>
+        write(ESET) *> writeSeq(eSet.sortedPs, writePar) *> writeOpt(eSet.remainder, writePar)
```

Comment:
Can list and set be in one line as before?

Reply:
writeSeq is now parametrised over type so accepts one more argument. I can only do this if remove this argument somehow. But I don't want to pull ParN type into the PrimitiveWriterOps
```scala
  def writeSeq[T](seq: Seq[T], writeT: T => F[Unit])(
      implicit applicativeF: Applicative[F]
  ): F[Unit] =
    writer.write(seq.size) *> seq.traverse_(writeT)
```